Certs names should be taken dynamically in indexer-security-init.sh

miguelfdez99 opened the pull request:
|Related issue|
|---|
|#1409|

## Description
Certificates names were not taken dynamically, this had to be changed.

In order to change this problem, the following things had been done:
- wazuh_indexer_root_ca is a variable which take root-ca.pem from the file opensearch.yml
-  WAZUH_INDEXER_ADMIN_PATH contains the directory where root-ca is stored.
- If admin.pem and admin-key.pem are in the same directory as root-ca it uses them.
- If not it will launch an error message giving the command to generate the keys.

## Logs example
I'm testing different use cases:
- 1: admin.pem does not exist
![1](https://user-images.githubusercontent.com/61159728/162734251-659105e9-6c17-4bbb-9252-b56e33e5870d.png)

- 2: admin-key.pem does not exist
![2](https://user-images.githubusercontent.com/61159728/162734405-d850010f-e5c4-44a7-9390-88747c81b150.png)

- 3: root-ca.pem does not exist
![3](https://user-images.githubusercontent.com/61159728/162734547-c51a2bdb-efe1-4867-a87c-afc87f662461.png)

- 4: wazuh-indexer it's not up
![4](https://user-images.githubusercontent.com/61159728/162735365-b73f0951-7b84-4dd9-9e80-469b13d3a619.png)

- 5: Using port 9650 instead of 9300
![5](https://user-images.githubusercontent.com/61159728/162735452-d565739b-abcf-49cc-a1d6-7c56d5dfce9d.png)

- 6: network.host it's 0.0.0.0
![6](https://user-images.githubusercontent.com/61159728/162735669-b10a5f6f-9286-48e4-8455-46902db2cdf3.png)

- 7: network.host it's a normal internal ip
Random private IP
![7-1](https://user-images.githubusercontent.com/61159728/162735883-77ca135f-e4db-41b6-9887-a8b64b01aefc.png)
IP from the other node
![7](https://user-images.githubusercontent.com/61159728/162736033-a29cee8d-c7e2-4a88-9400-a549c05780ef.png)

- 8: network.host name it's an interface 
![8](https://user-images.githubusercontent.com/61159728/162736217-bf8b535c-f929-44fa-8d81-8fa8a5bdbe33.png)

 9: /etc/wazuh-indexer/opensearch.yml does not exist
![9](https://user-images.githubusercontent.com/61159728/162736326-928c4de5-0355-4086-b92e-37e20c3dd9bb.png)

- 10: SECURITY_PATH does not exist
![10](https://user-images.githubusercontent.com/61159728/162736504-c094926f-9d5e-421e-ad62-7080a852ad85.png)

- 11: INSTALL_PATH does not exist
![11](https://user-images.githubusercontent.com/61159728/162736636-02b07090-a22e-4134-bed5-7cc567bab9f7.png)

- 12: CONFIG_PATH does not exist
![12](https://user-images.githubusercontent.com/61159728/162736705-27e3852a-db97-4025-ada5-9e205edeede5.png)

- 13: JAVA_HOME="${INSTALL_PATH}/jdk does not exist
It does not show an error but neither that it worked.
![13](https://user-images.githubusercontent.com/61159728/162736831-cae8ccde-a00a-4d69-af52-2284533a902f.png)

okynos commented on stack/indexer/indexer-security-init.sh:
```diff
@@ -42,11 +42,17 @@ fi
 
-OPENSEARCH_PATH_CONF="${CONFIG_PATH}" JAVA_HOME="${INSTALL_PATH}/jdk" runuser wazuh-indexer --shell="/bin/bash" --command="${SECURITY_PATH}/tools/securityadmin.sh -cd ${SECURITY_PATH}/securityconfig -cacert ${CONFIG_PATH}/certs/root-ca.pem -cert ${CONFIG_PATH}/certs/admin.pem -key ${CONFIG_PATH}/certs/admin-key.pem -h ${HOST} -p ${PORT} ${OPTIONS}"
-
+if [ -f "${WAZUH_INDEXER_ADMIN_PATH}/admin.pem" ] && [ -f "${WAZUH_INDEXER_ADMIN_PATH}/admin-key.pem" ]; then
```
Check also if root CA is present.

miguelfdez99 replied:
Done! https://github.com/wazuh/wazuh-packages/commit/80727a55d4c3d650f180e3c67492d579c147a5f8

okynos commented on stack/indexer/indexer-security-init.sh:
```diff
@@ -42,11 +42,17 @@ fi
 
 # -----------------------------------------------------------------------------
 
+eval "wazuh_indexer_root_ca=( $(cat /etc/wazuh-indexer/opensearch.yml | grep plugins.security.ssl.transport.pemtrustedcas_filepath | sed 's/plugins.security.ssl.transport.pemtrustedcas_filepath:\s//') )"
```
```suggestion
wazuh_indexer_root_ca=$(cat /etc/wazuh-indexer/opensearch.yml | grep http.pemtrustedcas | sed 's/.*: //')"
```

miguelfdez99 replied:
Done! https://github.com/wazuh/wazuh-packages/commit/80727a55d4c3d650f180e3c67492d579c147a5f8

okynos reviewed (changes requested):
Only a few changes 👍🏼

okynos reviewed (approved):
LGTM

DFolchA commented on stack/indexer/indexer-security-init.sh:
```diff
@@ -42,11 +42,17 @@ fi
 # -----------------------------------------------------------------------------
 
+WAZUH_INDEXER_ROOT_CA=$(cat /etc/wazuh-indexer/opensearch.yml | grep http.pemtrustedcas | sed 's/.*: //')"
+WAZUH_INDEXER_ADMIN_PATH="${WAZUH_INDEXER_ROOT_CA%%/root-ca.*}"
```
What if the key is not named root-ca??
you can use `dirname $file` to obtain the path to the directory

miguelfdez99 replied:
Done https://github.com/wazuh/wazuh-packages/commit/59c81ff61ec2d872a466910d634b90ccef023e6a

alberpilot reviewed (changes requested):
- The `securityadmin` function has a wrong indentation
- The `SECURITY_PATH` variable can be placed after OPTIONS at the top. 
- It's needed to ensure that all checks output are null redirected. If not, this could be an output:
```
grep: /etc/wazuh-indexer/opensearch.yml: No such file or directory
grep: /etc/wazuh-indexer/opensearch.yml: No such file or directory
cat: /etc/wazuh-indexer/opensearch.yml: No such file or directory
dirname: missing operand
Try 'dirname --help' for more information.
ERROR: admin.pem and admin-key.pem could not be found in . You must execute the following command: JAVA_HOME=/usr/share/wazuh-indexer/jdk runuser wazuh-indexer --shell=/bin/bash --command=/usr/share/wazuh-indexer/plugins/opensearch-security/tools/securityadmin.sh -cd /usr/share/wazuh-indexer/plugins/opensearch-security/plugins/opensearch-security/securityconfig -cacert  -cert $/etc/wazuh-indexer/certs/admin.pem -key /etc/wazuh-indexer/certs/admin-key.pem -h  -p 9300 -icl -nhnv 
```
If some check failed, I only want to see the error message. 
- The sentence must be re-phrased by: ERROR: this tool try to find admin.pem and admin-key.pem in ${WAZUH_INDEXER_ADMIN_PATH} but it couldn't. In this case, you must run manually the Indexer security initializer by running the command: JAVA_HOME="/usr/share/wazuh-indexer/jdk" runuser wazuh-indexer --shell="/bin/bash" --command="/usr/share/wazuh-indexer/plugins/opensearch-security/tools/securityadmin.sh -cd /usr/share/wazuh-indexer/plugins/opensearch-security/plugins/opensearch-security/securityconfig -cacert /path/to/root-ca.pem -cert /path/to/admin.pem -key /path/to/admin-key.pem -h ${HOST} -p ${PORT} ${OPTIONS}" replacing `/path/to/` by your certificates path."
- Why `$` in `/etc/wazuh-indexer/certs/admin.pem` path?

alberpilot reviewed (changes requested):
Based on your provided user case testing:
1. Message expected. :heavy_check_mark: 
2. Message expected. :heavy_check_mark: 
3. Message expected. :heavy_check_mark: 
4. Message expected. :heavy_check_mark: 
5. Message expected. :heavy_check_mark: 
6. If network.host = 0.0.0.0, the script should use 127.0.0.1, as you can see in the code, getNetworkHost function. Please investigate why. :red_circle: 
7. Message expected. :heavy_check_mark: 
8. If network.host = eth0, the script should use the IP associated to the interface, as getNetworkFunction looks to be defining. Please investigate why. :red_circle: 
9. If `/etc/wazuh-indexer/opensearch.yml` doesn't exist, the script must show a message saying: `it was not possible to find /etc/wazuh-indexer/opensearch.yml`. :red_circle: 
10. Same. :red_circle: 
11. Same. :red_circle: 
12. The script should be able to control de error message. :red_circle: 
13. The script should be able to show an error message if JAVA_HOME doesn't exist. :red_circle: 
14.

miguelfdez99 commented:
- 6:  network.host it's 0.0.0.0
![6](https://user-images.githubusercontent.com/61159728/162942468-0eb2ea35-4c9c-4c75-9d38-874c97b02c2d.png)

- 8: network.host name it's an interface
I've tried several IPs and interface names and everything works as expected. I've changed the function getNetworkHost() in order to achieve this. https://github.com/wazuh/wazuh-packages/pull/1441/commits/df46a66ea69a2530d7e3f6d0dc139d742f5e2beb
![8](https://user-images.githubusercontent.com/61159728/162942586-03ba511f-86fa-4f49-b962-536ea70f3f9d.png)

- https://github.com/wazuh/wazuh-packages/pull/1441/commits/74009526f390e74188438370ec57da99dece5ab9
In this commit I have added several error messages that could happen.

- 9: /etc/wazuh-indexer/opensearch.yml does not exist
![9](https://user-images.githubusercontent.com/61159728/162943488-a4cd56b7-3cb3-4bfa-a558-b6f1bde2e053.png)

- 10: SECURITY_PATH does not exist
![10](https://user-images.githubusercontent.com/61159728/162943777-243d37ab-9573-45f3-9e87-3a327010ed40.png)

- 11: INSTALL_PATH does not exist
![11](https://user-images.githubusercontent.com/61159728/162943861-ee4834bc-2809-4d99-b9c4-1f6708247665.png)

- 12: CONFIG_PATH does not exist
![12](https://user-images.githubusercontent.com/61159728/162943923-5c4e9519-9182-473a-9430-b7e1f59e7195.png)

- 13: JAVA_HOME="${INSTALL_PATH}/jdk does not exist
![13](https://user-images.githubusercontent.com/61159728/162944002-bf9ec7bd-1865-458d-89e8-fc9ca43edb47.png)

okynos commented on stack/indexer/indexer-security-init.sh:
```diff
@@ -12,40 +12,76 @@ HOST=""
+WAZUH_INDEXER_ROOT_CA="$(cat /etc/wazuh-indexer/opensearch.yml 2>&1 | grep http.pemtrustedcas | sed 's/.*: //')"
+WAZUH_INDEXER_ADMIN_PATH="$(dirname "${WAZUH_INDEXER_ROOT_CA}" 2>&1)"
+SECURITY_PATH="${INSTALL_PATH}/plugins/opensearch-security"
+CONFIG_FILE="${CONFIG_PATH}/opensearch.yml"
```
Set up the first and use in the rest of calls e.g. `L15`

okynos commented on stack/indexer/indexer-security-init.sh:
```diff
@@ -12,40 +12,76 @@ HOST=""
-OPENSEARCH_PATH_CONF="${CONFIG_PATH}" JAVA_HOME="${INSTALL_PATH}/jdk" runuser wazuh-indexer --shell="/bin/bash" --command="${SECURITY_PATH}/tools/securityadmin.sh -cd ${SECURITY_PATH}/securityconfig -cacert ${CONFIG_PATH}/certs/root-ca.pem -cert ${CONFIG_PATH}/certs/admin.pem -key ${CONFIG_PATH}/certs/admin-key.pem -h ${HOST} -p ${PORT} ${OPTIONS}"
+    getPort
+
+    if [ ! -f "${CONFIG_FILE}" ]; then
+        echo "ERROR: it was not possible to find ${CONFIG_FILE}"
+        exit 1
```
Move to the Line 10

miguelfdez99 replied:
Done https://github.com/wazuh/wazuh-packages/pull/1441/commits/c88e0779f55f2646faab03b81d4c9a9e5a7fa551

okynos reviewed (changes requested):
Minor changes

okynos commented on stack/indexer/indexer-security-init.sh:
```diff
@@ -8,44 +8,88 @@
 
-    if [ "${HOST}" = "0.0.0.0" ]; then
-        HOST="127.0.0.1"
+    if [ -f "${CONFIG_FILE}" ]; then   
```
Not needed

okynos commented on stack/indexer/indexer-security-init.sh:
```diff
@@ -8,44 +8,88 @@
 # -----------------------------------------------------------------------------
+getPort() {
+
+    if [ -f "${CONFIG_FILE}" ]; then
```
Not needed

okynos reviewed (changes requested):
Minor change, GJ

okynos reviewed (approved):
LGTM
